pinot-broker/src/main/java/org/apache/pinot/broker/routing/BrokerRoutingManager.java

```diff
@@ -744,4 +769,63 @@ InstanceSelector.SelectionResult calculateRouting(BrokerRequest brokerRequest, l
       }
     }
   }
+
+  private void buildTableTenantServerMap(String tableNameWithType, TableConfig tableConfig) {
+    LOGGER.info("buildTableTenantServerMap");
+    String serverTag = getServerTagForTable(tableNameWithType, tableConfig);
+    List<InstanceConfig> allInstanceConfigs = HelixHelper.getInstanceConfigs(_helixManager);
+    List<InstanceConfig> instanceConfigsWithTag = HelixHelper.getInstancesConfigsWithTag(allInstanceConfigs, serverTag);
+    Map<String, ServerInstance> serverInstances = new HashMap<>();
+    for (InstanceConfig serverInstanceConfig : instanceConfigsWithTag) {
+      LOGGER.info("Building instances. table={}, host={}, instance={}", tableNameWithType,
+          serverInstanceConfig.getHostName(), serverInstanceConfig.getInstanceName());
+      serverInstances.put(serverInstanceConfig.getInstanceName(), new ServerInstance(serverInstanceConfig));
+    }
+    _tableTenantServersMap.put(tableNameWithType, serverInstances);
+  }
+
+  private void addNewServerToTableTenantServerMap(String instanceId, ServerInstance serverInstance,
+      InstanceConfig instanceConfig) {
+    LOGGER.info("addNewServerToTableTenantServerMap");
+    List<String> tags = instanceConfig.getTags();
+
+    for (Map.Entry<String, Map<String, ServerInstance>> entry : _tableTenantServersMap.entrySet()) {
+      String tableNameWithType = entry.getKey();
+      TableConfig tableConfig = ZKMetadataProvider.getTableConfig(_propertyStore, tableNameWithType);
+      String tableServerTag = getServerTagForTable(tableNameWithType, tableConfig);
+
+      Map<String, ServerInstance> tenantServerMap = entry.getValue();
+
+      LOGGER.info("Iterating for table={}, tag={}, server_instance_id={}, host={}", tableNameWithType, tableServerTag,
+          instanceId, serverInstance.getHostname());
+
+      if (tags.contains(tableServerTag) && !tenantServerMap.containsKey(instanceId)) {
```

Comment:
May be `tags` should be converted to a `Set` ?

Reply:
Should the 2nd check be the first check ?

Reply:
Changed order of conditions. I think having a list is fine since that's how helix function returns? Highly unlikely that an instance is configured with duplicate tags. And correctness wise, it shouldn't be a problem.